src/lib/zf/log.c

```diff
@@ -137,12 +138,19 @@ void zf_log(struct zf_stack* st, const char* fmt, ...)
 
 int zf_log_redirect(const char* file)
 {
-  FILE* f = fopen(file, "a");
+  strncpy(zf_log_file_name, file, sizeof(zf_log_file_name));
+  zf_log_file_name[ZF_LOG_FILE_NAME_SIZE-1] = '\0';
+
+  if( sizeof(file) > ZF_LOG_FILE_NAME_SIZE-1 ) 
```

Comment:
When you declare an array like `char a[10];`, the name `a` represents the whole array, not just a pointer to the first element (though it can also be used as such). `sizeof(a) == 10` in this case, and that was my intention in the code above. Moreover, the next line of code ensures that the destination string is always null-terminated (as `strncpy` does not guarantee so).

Reply:
File is not of array type in this case, it's a pointer so `sizeof(file) ==8` in this case

Reply:
Indeed, with a bit more testing, it also seems that changing `file` to be `char[]` still wouldn't be safe (as in this case I think it's implicitly cast to a `char*` when passed). See this [small snippet I used to test with godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:1,endLineNumber:19,positionColumn:1,positionLineNumber:19,selectionStartColumn:1,selectionStartLineNumber:19,startColumn:1,startLineNumber:19),source:'%23include+%3Cstdio.h%3E%0A%0Asize_t+string_length(char*+str)+%7B%0A++return+sizeof(str)%3B%0A%7D%0A%0Asize_t+buf_length(char+str%5B%5D)+%7B%0A++return+sizeof(str)%3B%0A%7D%0A%0Aint+main(void)%0A%7B%0A++char+buf%5B512%5D+%3D+%22/dev/stderr%22%3B%0A%0A++printf(%22indirect+sizeof(char*)+%3D+%25ld%5Cn%22,+string_length(buf))%3B%0A++printf(%22direct+sizeof(char*)+%3D+%25ld%5Cn%22,+sizeof((char*)buf))%3B%0A++printf(%22indirect+sizeof(char%5B%5D)+%3D+%25ld%5Cn%22,+buf_length(buf))%3B%0A++printf(%22direct+sizeof(char%5B%5D)+%3D+%25ld%5Cn%22,+sizeof(buf))%3B%0A%0A++return+0%3B%0A%7D'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:49.50101537568901,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:cclang_trunk,compilerName:'',compilerOutShown:'1',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'',overrides:!(),runtimeTools:!(),source:1,stdinPanelShown:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+clang+(trunk)+(C,+Editor+%231)',t:'0'),(h:compiler,i:(compiler:cg141,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.1+(Editor+%231)',t:'0')),header:(),k:50.498984624311,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Reply:
yes, you're correct. Sorry for missing this.